src/main/java/org/rocstreaming/roctoolkit/NativeObjectReference.java

```diff
@@ -0,0 +1,103 @@
+package org.rocstreaming.roctoolkit;
+
+import java.lang.ref.PhantomReference;
+import java.lang.ref.ReferenceQueue;
+import java.util.Optional;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicReference;
+
+/**
+ * <code>NativeObjectReference</code> is associated with a {@link NativeObject} and owns its entire lifetime;
+ *
+ * A <code>NativeObjectReference</code> contains necessary data for closing the native object
+ * after it becomes phantom reachable.
+ */
+class NativeObjectReference extends PhantomReference<NativeObject> implements AutoCloseable {
```

Comment:
Could you clarify why we need AutoCloseable here?

Reply:
Here it is just to explicitly expose a `close` method by that interface.